Tests/Common/SimulatedDevice.cs

```diff
@@ -129,15 +125,27 @@ public LoRaPayloadData CreateUnconfirmedDataUpMessage(string data, uint? fcnt =
                 direction,
                 Supports32BitFCnt ? fcnt : null);
 
+            if (AppSKey is null || NwkSKey is null) throw new InvalidOperationException("Can't perform encryption as some keys are empty");
+            payloadData.PerformEncryption(string.IsNullOrEmpty(appSKey) ? AppSKey : appSKey);
+            payloadData.SetMic(string.IsNullOrEmpty(nwkSKey) ? NwkSKey : nwkSKey);
+
+            // We want to ensure we simulate a message coming from the device, therefore only the 16 bits of the framecounter should be available.
+            // The following line ensure we remove the 32 bits of the server frame counter that was generated by the constructor.
+            // Some tests cases are expecting the mic check to fail because of rollover of the fcnt, if we have this value set it will never fail.
+            payloadData.Reset32BitBlockInfo();
+            // THIS IS NEEDED FOR TESTS AS THIS CONSTRUCTOR IS USED THERE
+            // THIS WILL BE REMOVED WHEN WE MIGRATE TO USE lORAPAYLOADDATALNS INSTEAD OF LORAPAYLOAD in #1085
+            // Populate the MacCommands present in the payload
+            if (payloadData.Fopts.Length != 0)
+                payloadData.MacCommands = MacCommand.CreateMacCommandFromBytes(payloadData.Fopts);
+
             return payloadData;
         }
 
-        public UplinkPktFwdMessage CreateConfirmedMessageUplink(string data, uint? fcnt = null, byte fport = 1) => CreateConfirmedDataUpMessage(data, fcnt, fport).SerializeUplink(AppSKey, NwkSKey);
-
         /// <summary>
         /// Creates request to send unconfirmed data message.
         /// </summary>
-        public LoRaPayloadData CreateConfirmedDataUpMessage(string data, uint? fcnt = null, byte fport = 1, bool isHexPayload = false)
+        public LoRaPayloadData CreateConfirmedDataUpMessage(string data, uint? fcnt = null, byte fport = 1, bool isHexPayload = false, string appSKey = "", string nwkSKey = "")
```

Comment:
Why default `""` for `appSKey` and `nwkSKey` here instead of `null`? Seems inconsistent with the other overload?

Reply:
`CreateUnconfirmedDataUpMessage` and `CreateConfirmedDataUpMessage` seem to have a lot of shared code. Can the duplication not be consolidated so we don't have to maintain or make fixes twice?

Reply:
I harmonized all to be null

Reply:
If that is fine with you I'd prefer tackle the merging in a subsequent PR as I have no idea if I would break some tests, created #1109